src/LagrangianParticleTracking.jl

```diff
@@ -0,0 +1,84 @@
+module LagrangianParticleTracking
+
+export LagrangianParticles, advect_particles!
+
+using KernelAbstractions
+
+using Oceananigans.Grids
+
+using Oceananigans.Architectures: device
+using Oceananigans.Fields: interpolate, datatuple
+using Oceananigans.Utils: MAX_THREADS_PER_BLOCK
+
+import Base: size, length
+
+struct LagrangianParticles{X, Y, Z}
+    x :: X
+    y :: Y
+    z :: Z
+end
```

Comment:
I haven't worked out all the details, but I wonder if abstracting `model.particles` as an `Array` or `CuArray` of `AbstractParticle`s is a better design. Consider:

```julia
struct ReactingTracerParticle{P} <: AbstractParticle
    position :: P
    species :: P
end
```

We can then provide an interface for defining reactions and such things on particles.

This use is futuristic, but I don't think we lose anything by modifying the design to anticipate uses like this. It's also simpler because we don't have to check that `x, y, z` are the same length, etc.

Reply:
Hmm, perhaps what we would want to support is a type that can extract field values at the particle position.

Then we can design an interface for writing RHS that depend on the field values so extracted, so that reactions-on-particles can depend arbitrarily on tracer/velocity values at the particle location.

This more advanced type doesn't have to be part of this PR but I think we should set off on the right foot.

Reply:
Not sure you know about `StructArrays.jl`. I'm using it now for particles in `PlanktonIndividuals.jl`. It works on both CPU and GPU.

Reply:
That's interesting --- but actually wouldn't it be more efficient to keep each position `x, y, z` close in memory? For example, @ali-ramadhan's kernel iterates over the list of particles, rather than each coordinate individually. This is better served by the layout

`[(x1, y1, z1), (x2, y1, z2), ...]`

(though in practice it may not matter much).

Reply:
`StructArrays.jl` could definitely prove useful in some capacity so thanks for pointing it out!

Reply:
Yes, you're right. The reason I use `StructArrays.jl` is that it is easy to broadcast at `x, y, z`. But if you use `@kernel` functions, the first one is more readable.